packages/loopring_v3/contracts/amm/libamm/AmmTransactionReceiver.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2017 Loopring Technology Limited.
+pragma solidity ^0.7.0;
+pragma experimental ABIEncoderV2;
+
+import "../../core/impl/libtransactions/BlockReader.sol";
+import "../../lib/MathUint.sol";
+import "./AmmData.sol";
+import "./AmmExitProcess.sol";
+import "./AmmJoinProcess.sol";
+import "./AmmPoolToken.sol";
+import "./AmmUpdateProcess.sol";
+
+
+/// @title AmmTransactionReceiver
+library AmmTransactionReceiver
+{
+    using AmmExitProcess    for AmmData.State;
+    using AmmJoinProcess    for AmmData.State;
+    using AmmPoolToken      for AmmData.State;
+    using AmmUpdateProcess  for AmmData.Context;
+    using BlockReader       for bytes;
+
+    function onReceiveTransactions(
+        AmmData.State storage  S,
+        bytes         calldata txsData,
+        bytes         calldata callbackData
+        )
+        internal
+    {
+        AmmData.Context memory ctx = _getContext(S, txsData);
+
+        ctx.approveAmmUpdates();
+
+        _processPoolTx(S, ctx, callbackData);
+
+        // Update state
+        S._totalSupply = ctx.totalSupply;
+
+        // Make sure we have consumed exactly the expected number of transactions
+        require(txsData.length == ctx.txsDataPtr - ctx.txsDataPtrStart, "INVALID_NUM_TXS");
+    }
+
+    function _getContext(
+        AmmData.State storage  S,
+        bytes         calldata txsData
+        )
+        private
+        view
+        returns (AmmData.Context memory)
+    {
+        uint size = S.tokens.length;
+        uint txsDataPtr = 23;
+        assembly {
+            txsDataPtr := sub(add(txsData.offset, txsDataPtr), 32)
+        }
+        return AmmData.Context({
+            txsDataPtr: txsDataPtr,
+            txsDataPtrStart: txsDataPtr,
+            domainSeparator: S.domainSeparator,
+            accountID: S.accountID,
+            poolTokenID: S.poolTokenID,
+            feeBips: S.feeBips,
+            totalSupply: S._totalSupply,
+            tokens: S.tokens,
+            tokenBalancesL2: new uint96[](size)
+        });
+    }
+
+    function _processPoolTx(
+        AmmData.State   storage  S,
+        AmmData.Context memory   ctx,
+        bytes           calldata callbackData
+        )
+        private
+    {
+        // Manually decode the encoded PoolTx in `callbackData`
```

Comment:
Yes, no other reason. This saves on an abi decode that copies the data from calldata to memory. And that would also make the 2nd abi decode that is still used for decoding the PoolJoin/PoolExit more expensive because it would then need to be done from memory.